toolchain/check/import_ref.cpp

```diff
@@ -885,27 +903,40 @@ class ImportRefResolver {
   // before, in which case it's the previous result.
   //
   // TODO: Error is returned when support is missing, but that should go away.
-  auto TryResolveInst(SemIR::InstId inst_id, SemIR::ConstantId const_id)
-      -> ResolveResult {
+  auto TryResolveInst(SemIR::InstId inst_id, SemIR::ConstantId const_id,
+                      SemIR::ConstantId abstract_const_id) -> ResolveResult {
     auto inst_const_id = import_ir_.constant_values().Get(inst_id);
     if (!inst_const_id.is_valid() || !inst_const_id.is_symbolic() ||
         const_id.is_valid()) {
+      CARBON_CHECK(!abstract_const_id.is_valid())
+          << "Should not have an abstract constant from a previous pass when "
+             "not resolving a symbolic constant";
       return TryResolveInstCanonical(inst_id, const_id);
     }
 
-    // Try to import the generic, and retry if it's not ready yet. Note that if
-    // this retries, we can require three passes to import an instruction.
+    // Try to import the generic. This might add new work.
     const auto& symbolic_const =
         import_ir_.constant_values().GetSymbolicConstant(inst_const_id);
-    auto initial_work = work_stack_.size();
     auto generic_const_id = GetLocalConstantId(symbolic_const.generic_id);
-    if (HasNewWork(initial_work)) {
-      return ResolveResult::Retry();
-    }
+    bool need_to_import_generic = HasNewWork();
 
     // Import the constant and rebuild the symbolic constant data.
-    auto result = TryResolveInstCanonical(inst_id, const_id);
-    if (result.const_id.is_valid()) {
+    auto result = abstract_const_id.is_valid()
+                      ? ResolveResult{.const_id = abstract_const_id}
+                      : TryResolveInstCanonical(inst_id, const_id);
+
+    if (need_to_import_generic) {
+      if (result.const_id.is_valid()) {
+        CARBON_CHECK(result.state == ResolveResult::Finished)
+            << "Should not retry with a constant result with if we already "
+               "added new work";
```

Comment:
I'm having trouble reading this, "with ... with if"

Reply:
Resolving, this code no longer exists.